### ISSUE TITLE
Intelligently handle automatically leaving presence in `usePresence` hook on component unmount

### ISSUE THREAD

**Issue by VeskeR** [ably/ably-js#1687]
TLDR: current issue is that if you have two components mounted with `usePresence` hook and one of them is umounted then you'll leave presence even though you still have a component with `usePresence` mounted.
For more details see discussion https://github.com/ably/ably-js/pull/1674#discussion_r1518012408.

Possible ideas on how to solve it:

1. Implement a ref counter for `usePresence` hooks. This way only last unmounted hook will actually leave presence. Can be done with something like:
```tsx
// presenceCounter object in context in ChannelProvider
      _channelNameToChannelContext: {
        ..._channelNameToChannelContext,
        [channelName]: {
          channel,
          derived,
          presenceCounter: {}
        },
      },
```
and then incrementing/decrementing this counter in `usePresence` hook.
2. Let user handle entering and leaving presence on their own:
```tsx
const { enter, leave } = usePresence('channel-name');

useEffect(() => {
  enter();

  return () => {
    leave();
  };
}, [enter, leave]);
```
This basically just puts the responsibility to handle presence leaving logic on the user. Not ideal solution, it would be preferable if `usePresence` hook worked out of the box without the need to write a specific logic for it on user side.
3. 'PresenceProvider' component, but it would be less intuitive to use, there is nothing that is actually "provided" (presence is part of the channel, that is already provided by the `ChannelProvider`), and unclear how users would prevent it from entering presence set if they want to delay it.

Ref counter seems like a better choice.
